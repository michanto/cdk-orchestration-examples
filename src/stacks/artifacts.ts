import { AppToken, BUILD_TIME, Log, StackToken } from '@michanto/cdk-orchestration';
import { S3FileResource } from '@michanto/cdk-orchestration/orchestration';
import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Function, Runtime, Code } from 'aws-cdk-lib/aws-lambda';
import { IBucket, Bucket } from 'aws-cdk-lib/aws-s3';
import { md5hash } from 'aws-cdk-lib/core/lib/helpers-internal';
import { PhysicalResourceId } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

const NAMESPACE = '@michanto/cdk-orchestration-examples';

/** Token names for use with AppToken and StackToken . */
const NAMED_BUCKET_TKN = `${NAMESPACE}.NamedBucket`;
const UNNAMED_BUCKET_TKN = `${NAMESPACE}.UnnamedBucket`;
const BUILD_ID_TKN = `${NAMESPACE}.BuildId`;

/**
 * Class that creates the tokens we use in our App.
 */
export class SampleTokens {
  /** Resolves to the name of the Named Bucket. */
  static namedBucketNameToken(scope: Construct) {
    return AppToken.string(scope, NAMED_BUCKET_TKN, { displayHint: 'NmdBkt' });
  }

  /**
   * Separate token creation from token resolution.
   * Useful for cross-stack references without having to pass constructs.
   */
  static resolveNamedBucketNameToken(scope: Construct) {
    AppToken.resolveString(scope, NAMED_BUCKET_TKN, {
      produce: () => {
        let stack = Stack.of(scope);
        return `artifact-bucket-${stack.account}-${stack.region}-${
          md5hash(stack.node.path).slice(0, 8).toLowerCase()
        }`;
      },
    });
  }

  /** Resolves to the name of the unnamed Bucket (name will be generated at deployment time). */
  static unnamedBucketNameToken(scope: Construct) {
    return AppToken.string(scope, UNNAMED_BUCKET_TKN, { displayHint: 'UnnmdBkt' });
  }

  /*
  * Separate token creation from token resolution.
  * Useful for cross-stack references without having to pass constructs.
  */
  static resolveUnnamedBucketNameToken(scope: Construct, bucket: IBucket) {
    AppToken.resolveString(scope, UNNAMED_BUCKET_TKN, {
      produce: () => {
        return bucket.bucketName;
      },
    });
  }

  /**
   * Token for the buildId, which defaults to a timestamp, but is configurable.
   *
   * At Synthesis time, this token is resolved by checking a dictionary attached to the
   * App construct.
   *
   * @param scope Any scope in the app will do, as this is an AppToken.
   * @returns A string that prints as `${Token[BLDID.<token id>]}`
   */
  static buildIdToken(scope: Construct) {
    return AppToken.string(scope, BUILD_ID_TKN, {
      displayHint: 'BLDID',
    });
  }

  /**
   * Resolves the token named "CdkOrche::BuildId"
   * Resolved to any, but CDK will call toString() when writing the value since the token
   * is a string. Note BUILD_TIME is a number.
   * @param scope Any scope in the app will do, as this is an AppToken.
   * @param value
   */
  static resolveBuildIdToken(scope: Construct, value: any = BUILD_TIME) {
    AppToken.resolveAny(scope, BUILD_ID_TKN, { produce: () => value });
  }
}


export class OtherStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactBucketsStack', props?: StackProps) {
    super(scope, id, props);
    let log = Log.of(this);

    log.info('// @michanto/cdk-orchestration AppToken //');
    let buildIdToken = AppToken.string(this, 'build_id', {
      displayHint: 'BLDID',
    });
    log.info(JSON.stringify(buildIdToken));
    log.info(this.toJsonString(buildIdToken));

    // Application scoped named token.
    let bucketToken = AppToken.string(this, 'code_bucket', {
      displayHint: 'Bkt',
    });

    log.info('// @michanto/cdk-orchestration StackToken //');
    let app = this.node.root as App;

    let valinor = new Stack(app, `${this.stackName}-Valinor`, {
      env: {
        account: this.account, region: 'us-east-1',
      },
    });

    let beleriand = new Stack(app, `${this.stackName}-Beleriand`, {
      env: {
        account: this.account, region: 'eu-west-1',
      },
    });
    log.info('// Create Stack token in two different stacks. //');
    for (let stack of [valinor, beleriand]) {
      let keyToken = StackToken.string(stack, 'code_key', {
        displayHint: 'Key',
      });
      let codeBucket = Bucket.fromBucketName(stack, 'CodeBucket', bucketToken);
      new Function(stack, 'MyFunction', {
        runtime: Runtime.NODEJS_16_X,
        code: Code.fromBucket(codeBucket, keyToken),
        handler: 'handler',
      });
      StackToken.resolveString(stack, 'code_key', {
        produce: () => `${
          stack.stackName.toLowerCase()
        }/${
          buildIdToken
        }/code.zip`,
      });
    }
    let unresolvedToken = StackToken.string(this, 'unresolved_token', {
      displayHint: 'Unres',
    });
    log.info(unresolvedToken);
    try {
      log.info(`UnresolvedToken = '${this.resolve(unresolvedToken)}'`);
    } catch {
      log.info(`Could not resolve ${unresolvedToken}`);
    }

    StackToken.resolveString(this, 'unresolved_token', { produce: () => 'Now Resolved' });
    try {
      log.info(`UnresolvedToken = '${this.resolve(unresolvedToken)}'`);
    } catch {
      log.info(`Could not resolve ${unresolvedToken}`);
    }

    // Resolve app tokens.
    AppToken.resolveAny(this, 'build_id', { produce: () => BUILD_TIME });
    // Explicitly named bucket for use cross-region.
    let codeBucketName = `code-bucket-${this.account}`;
    new Bucket(this, 'CodeBucket', {
      bucketName: codeBucketName,
    });
    // Resolve bucket name.
    AppToken.resolveString(this, 'code_bucket', { produce: () => codeBucketName });

    log.info(this.resolve(buildIdToken));
  }
}
/**
 * The artifact bucket will be used to store artifact creation products.
 */
export class ArtifactBucketsStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactBucketsStack', props?: StackProps) {
    super(scope, id, props);

    new Bucket(this, 'NamedBucket', {
      bucketName: SampleTokens.namedBucketNameToken(this),
    });
    SampleTokens.resolveNamedBucketNameToken(this);

    const unnamedBucket = new Bucket(this, 'UnnamedBucket', {
    });
    SampleTokens.resolveUnnamedBucketNameToken(this, unnamedBucket);
  }
}

/**
 * Rather than pass the ArtifactBucket to this class, it can use the token.
 */
export class ArtifactWriterStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactWriter', props?: StackProps) {
    super(scope, id, props);
    let namedBucket = Bucket.fromBucketName(this, 'NamedBucket', SampleTokens.namedBucketNameToken(this));
    let unnamedBucket = Bucket.fromBucketName(this, 'UnnamedBucket', SampleTokens.unnamedBucketNameToken(this));

    new S3FileResource(this, 'Written2NamedBucket', {
      bucket: namedBucket,
      key: 'buildid.json',
      body: {
        buildId: SampleTokens.buildIdToken(this),
      },
      physicalResourceId: PhysicalResourceId.of('buildid.json'),
    });
    new S3FileResource(this, 'Written2UnnamedBucket', {
      bucket: unnamedBucket,
      key: 'buildid.json',
      body: {
        buildId: SampleTokens.buildIdToken(this),
      },
      physicalResourceId: PhysicalResourceId.of('buildid.json'),
    });
  }
}
