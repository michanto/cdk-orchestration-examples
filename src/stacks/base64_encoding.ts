import { Logger, LogLevel } from '@michanto/cdk-orchestration';
import { CustomResourceUtilities } from '@michanto/cdk-orchestration/custom-resources';
import { Echo, ImportOrders } from '@michanto/cdk-orchestration/transforms';
import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { GreetingLambdaTask } from '../constructs/greeting_lambda_task';

/**
 * Echo is a Transform, and Transforms work out-of-the-box with L1 or L2 constructs.
 * Otherwise, the default is for a transform to be applied to the stack.
 *
 * Since we want this Transform to be applied to the custom resource in an L3 construct
 * (such as Task or AwsCustomResource), we need to override the target property to
 * return the correct L1 construct.
 */
export class EchoCustomResource extends Echo {
  get target() {
    return new CustomResourceUtilities().findCustomResource(this.node.scope!);
  }
}

export class EchoCustomResourceAfterEncode extends EchoCustomResource {
  get order() {
    // Encode runs in WRITER order.
    return ImportOrders.WRITER;
  }
}

export class Base64EncodingExample extends Stack {
  constructor(scope: Construct, id: string = 'TransformsInfo', props?: StackProps) {
    super(scope, id, props);
    console.log('OrderedTransforms constructor BEGIN');
    Logger.set(this, new Logger({ logLevel: LogLevel.DEBUG }));

    /**
     * Similar to CDK Triggers, LambdaTask classes are custom resources that call a lambda.
     */
    let task = new GreetingLambdaTask(this, 'GreetingLambdaTask', {
      greeting: 'Hello, everyone!',
    });
    new EchoCustomResource(task, 'EchoCustomResource');
    new EchoCustomResourceAfterEncode(task, 'EchoCustomResourceAfterEncode');

    console.log('OrderedTransforms constructor END');
  }
}
