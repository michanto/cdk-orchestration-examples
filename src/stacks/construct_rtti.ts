import { ConstructRunTimeTypeInfo, ConstructTreeSearch, IStopCondition, Log, Logger, LogLevel } from '@michanto/cdk-orchestration';
import { App, CfnElement, Stack, StackProps } from 'aws-cdk-lib';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct, IConstruct } from 'constructs';
import { NAMESPACE } from '../private/internal';

/** A construct with RTTI for testing. */
export class TypedConstruct extends Construct {
  /** Type assertion for TypedConstruct. */
  static isTypedConstruct(x: IConstruct): x is TypedConstruct {
    return TypedConstruct.TYPED_CONSTRUCT_RTTI.hasRtti(x);
  }

  /**
   * Return the TypedConstruct hosting the given construct.
   */
  static of(x: IConstruct): TypedConstruct | undefined {
    let host = TypedConstruct.TYPED_CONSTRUCT_RTTI.searchUp(x);
    return host as TypedConstruct | undefined;
  }

  /**
   * Return all typed constructs under the given scope.
   */
  static typedConstructs(scope: IConstruct, stopCondition?: IStopCondition) {
    return ConstructTreeSearch.for(TypedConstruct.isTypedConstruct)
      .searchDown(scope, stopCondition);
  }

  private static readonly TYPED_CONSTRUCT_RTTI = new ConstructRunTimeTypeInfo({
    servicePropertyName: `${NAMESPACE}.test.TypedConstruct`,
  });

  constructor(scope: Construct, id: string) {
    super(scope, id);
    TypedConstruct.TYPED_CONSTRUCT_RTTI.addRtti(this);
  }
}

/**
 * Demonstrate built in and custom RTTI in the CDK.
 */
export class ConstructRtti extends Stack {
  constructor(scope: IConstruct, id: string, props?: StackProps) {
    super(scope, id, props);
    let app = this.node.root as App;
    Logger.set(app, new Logger({ logLevel: LogLevel.INFO }));
    let log = Log.of(this);

    // Built in RTTI
    log.info(`This construct is a stack: ${Stack.isStack(this)}`);
    log.info(`This construct is a construct: ${Construct.isConstruct(this)}`);
    log.info(`This construct is a CfnElement?: ${CfnElement.isCfnElement(this)}`);

    new TypedConstruct(this, 'UnderStack');
    new TypedConstruct(this.node.root, 'UnderApp');
    let bucket = new Bucket(this, 'MyBucket');
    new TypedConstruct(bucket, 'UnderBucket');
    new TypedConstruct(bucket.node.defaultChild!, 'UnderCfnBucket');
    let parent = new TypedConstruct(this, 'AParent');
    let hostedBucket = new Bucket(parent, 'HostedBucket');
    let subStack = new Stack(this, 'SubStack');
    new TypedConstruct(subStack, 'UnderSubStack');

    log.info(`isTypedConstruct true example: ${
      TypedConstruct.isTypedConstruct(parent)}`);
    log.info(`isTypedConstruct false example: ${
      TypedConstruct.isTypedConstruct(bucket)}`);
    log.info(`TypedConstructs in this app:\n${
      TypedConstruct.typedConstructs(app).map(c => c.node.path).join('\n  ')}`);
    log.info(`TypedConstructs in this stack:\n${
      TypedConstruct.typedConstructs(this, Stack.isStack).map(c => c.node.path).join('\n  ')}`);
    log.info(`TypedConstructs under this stack:\n${
      TypedConstruct.typedConstructs(this).map(c => c.node.path).join('\n  ')}`);
    log.info(`TypedConstructs in the sub stack:\n${
      TypedConstruct.typedConstructs(subStack).map(c => c.node.path).join('\n  ')}`);
    log.info(`TypedConstruct above hostedBucket:\n${
      TypedConstruct.of(hostedBucket)!.node.path}`);
    // Get the stack of all TypedConstructs under the Stack
    TypedConstruct.typedConstructs(this).forEach(t => Stack.of(t));
  }
}
