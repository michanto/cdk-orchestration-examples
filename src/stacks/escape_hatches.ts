import { BUILD_TIME, CfnElementUtilities, ConstructRunTimeTypeInfo, Log } from '@michanto/cdk-orchestration';
import { InlineNodejsFunction } from '@michanto/cdk-orchestration/aws-lambda-nodejs';
import { CustomResourceUtilities, RunResourceAlways } from '@michanto/cdk-orchestration/custom-resources';
import { CfnElement, CfnResource, Stack, StackProps } from 'aws-cdk-lib';
import { CfnRole } from 'aws-cdk-lib/aws-iam';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from 'aws-cdk-lib/custom-resources';
import { Trigger } from 'aws-cdk-lib/triggers';
import { Construct, IConstruct } from 'constructs';
import { NAMESPACE } from '../private/internal';

const LAMBDA_PATH = `${__dirname}/../../lib/constructs/lambdas/`;

export class StackDescription extends CfnElement {
  constructor(scope: Construct, id: string, readonly description: string) {
    super(scope, id);
  }

  _toCloudFormation(): object {
    return {
      Description: this.description,
    };
  }
}

/**
 * A well designed construct has a Properties interface with readonly members.
 */
export interface AddMetadataProperties {
  /** Metadata key. */
  readonly key: string;
  /** Metadata value. */
  readonly value: any;
}

/**
 * This construct adds metadata to a Stack or a CfnResource.
 *
 * A well designed construct has documentation.
 *
 * An well designed escape hatch construct works with Stacks (if applicable),
 * L1, L2 or L3 constructs without the user having to navigate
 * the construct tree.
 */
export class AddMetadata extends Construct {
  /** RTTI is optional depending on your use case. */
  static isAddMetadata(x: IConstruct): x is AddMetadata {
    return AddMetadata.ADD_METADATA_CONSTRUCT_RTTI.hasRtti(x);
  }

  private static readonly ADD_METADATA_CONSTRUCT_RTTI = new ConstructRunTimeTypeInfo({
    servicePropertyName: `${NAMESPACE}.AddMetadata`,
  });

  constructor(scope: Construct, id: string, readonly props: AddMetadataProperties) {
    super(scope, id);

    AddMetadata.ADD_METADATA_CONSTRUCT_RTTI.addRtti(this);
    this.target.addMetadata(props.key, props.value);
  }

  /**
   * By puttng the target in a getter, subclasses can easily override the target.
   * This expands the usefulness of our Construct.
   * For example, the user could target a specfic resource type in a sub tree.
   */
  get target(): CfnResource | Stack {
    let scope = this.node.scope!;

    // If this construct is added to a stack, the user wants to add stack metadata.
    // Otherwise, the user wants to add CfnResource metadata.
    if (Stack.isStack(scope)) {
      return scope;
    }

    // Otherwise the user wants to add metadata to a CfnResource under scope.
    return new CfnElementUtilities().findCfnResource(scope);
  }
}

/**
 * Adds Salt property to any custom resource.
 * NOTE:  This exists in cdk-orchestration as RunResourceAlways construct.
 *
 * In the CDK Triggers [documentation](
 * https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.triggers-readme.html) it says:
 * >> In the future we will consider adding support for additional re-execution modes:
 *
 * >> `executeOnEveryDeployment`: boolean - re-executes every time the stack is
 * deployed (add random "salt" during synthesis).
 *
 * We can go ahead and implement that, but for ANY custom resource.
 */
export class AddSalt extends Construct {
  constructor(scope: Construct, id: string = 'AddSalt', readonly saltValue: number = BUILD_TIME ) {
    super(scope, id);
    this.target.addPropertyOverride('salt', saltValue);
    Log.of(this).debug(() => `Added salt value ${saltValue}.`);
  }

  get target() {
    let scope = this.node.scope!;
    return new CustomResourceUtilities().findCustomResource(scope);
  }
}

/**
 * Show how to use [Escape Hatches](https://docs.aws.amazon.com/cdk/v2/guide/cfn_layer.html)
 * in the CDK.
 *
 * The 'cfn layer' document recommends using L1 methods and raw overrides as escape hatches.
 * Here, we'll see that we can encapsulate these as constructs and use searches to make working
 * with escape hatches more natural.
 */
export class EscapeHatches extends Stack {
  constructor(scope: Construct, id: string = 'WritingConstructs', props?: StackProps) {
    super(scope, id, { ...props, description: 'EscapeHatches description.' });
    let log = Log.of(this);

    new StackDescription(this, 'Description', 'This string will be in the stack description');
    new StackDescription(this, 'Description2', 'And so will this one.');

    new AddMetadata(this, 'StackMetadata', {
      key: 'Elf', value: 'Arondir',
    });

    let bucket = new Bucket(this, 'MyBucket');
    new AddMetadata(bucket, 'BucketMetadata', {
      key: 'Mountain', value: 'Taniquetil',
    });

    const echoLambda = new InlineNodejsFunction(this, 'EchoLambda', {
      entry: `${LAMBDA_PATH}/echo.js`,
      environment: {
        LogLevel: '1',
      },
    });

    new AddMetadata(echoLambda, 'FunctionMetadata', {
      key: 'Dwarf', value: 'Balin',
    });

    new class AddRoleMetadata extends AddMetadata {
      get target(): CfnResource | Stack {
        return new CfnElementUtilities().findCfnResource(this.node.scope!, CfnRole.CFN_RESOURCE_TYPE_NAME);
      }
    }(echoLambda, 'RoleMetadata', {
      key: 'Maia', value: 'Bombadillo',
    });

    let trigger = new Trigger(this, 'EchoTrigger', {
      handler: echoLambda,
    });
    log.debug(() => 'Echo lambda will trigger on every deployment.');
    new AddSalt(trigger);

    let writer = new AwsCustomResource(this, 'MyS3FileResource', {
      onUpdate: {
        service: 'S3',
        action: 'putObject',
        parameters: {
          Body: JSON.stringify({ dummy: 'data' }),
          Bucket: bucket.bucketName,
          Key: 'dummy.json',
        },
        physicalResourceId: PhysicalResourceId.of(`s3:://${bucket.bucketName}/dummy.json`),
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: [bucket.bucketArn, `${bucket.bucketArn}/*`],
      }),
    });
    new RunResourceAlways(writer);
  }
}
