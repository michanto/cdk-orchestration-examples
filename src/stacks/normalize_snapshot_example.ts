import { Echo } from '@michanto/cdk-orchestration/transforms';
import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { EnsureChangeInStack } from '../constructs/ensure_change_in_stack';
import { GreetingLambdaTask } from '../constructs/greeting_lambda_task';
import { NormalizeSnapshot } from '../constructs/normalize_snapshot';

export class NormalizeSnapshotExample extends Stack {
  constructor(scope: Construct, id: string = 'NormalizeSnapshotExample', props?: StackProps) {
    super(scope, id, props);
    EnsureChangeInStack.for(this);
    new GreetingLambdaTask(this, 'GreetingTaskLambda', {
      greeting: 'Hey come merry dol!',
    });

    new Echo(this, 'Unnormalized');
    new NormalizeSnapshot(this);
    new Echo(this, 'Normalized');
  }
}