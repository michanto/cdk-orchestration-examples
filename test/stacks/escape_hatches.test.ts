import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { EscapeHatches } from '../../src/stacks/escape_hatches';

test('WritingConstructs', () => {
  const app = new App();
  const stack = new EscapeHatches(app);

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();

});
