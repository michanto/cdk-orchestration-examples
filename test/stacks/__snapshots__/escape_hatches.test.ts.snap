// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EscapeHatches 1`] = `
{
  "Description": "EscapeHatches description.
This string will be in the stack description.
And so will this one.",
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Metadata": {
    "Elf": "Arondir",
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "ce2f3595a340d6c519a65888ef97e3b9b64f053f83608e32cc28162e22d7d99a.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91": {
      "DependsOn": [
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "fe27a81f1e8702187a8717117fe7516f45948616ac25cefe80cc7e202365594b.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          {
                            "Fn::GetAtt": [
                              "EchoLambda36664C86",
                              "Arn",
                            ],
                          },
                          ":*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EchoLambda36664C86": {
      "DependsOn": [
        "EchoLambdaServiceRole6F110512",
      ],
      "Metadata": {
        "Dwarf": "Balin",
      },
      "Properties": {
        "Code": {
          "ZipFile": ""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = handler;
function log(message) {
if (process.env.LogLevel) {
console.log(JSON.stringify(message));
}
}
async function handler(event, context) {
log({ Event: event });
log({ Context: context });
return Promise.resolve(event !== null && event !== void 0 ? event : {});
}",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "LogLevel": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "EchoLambdaServiceRole6F110512",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "EchoLambdaCurrentVersionC182B060d021074746858380712e465aa3dea2be": {
      "Properties": {
        "FunctionName": {
          "Ref": "EchoLambda36664C86",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "EchoLambdaServiceRole6F110512": {
      "Metadata": {
        "Maia": "Bombadillo",
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EchoTrigger": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ExecuteOnHandlerChange": true,
        "HandlerArn": {
          "Ref": "EchoLambdaCurrentVersionC182B060d021074746858380712e465aa3dea2be",
        },
        "InvocationType": "RequestResponse",
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
            "Arn",
          ],
        },
        "Timeout": "120000",
        "salt": 1730494729270,
      },
      "Type": "Custom::Trigger",
      "UpdateReplacePolicy": "Delete",
    },
    "MyBucketF68F3FF0": {
      "DeletionPolicy": "Retain",
      "Metadata": {
        "Mountain": "Taniquetil",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyS3FileResource1CAF1D27": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "MyS3FileResourceCustomResourcePolicy761A62A0",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"S3","action":"putObject","parameters":{"Body":"{\\"dummy\\":\\"data\\"}","Bucket":"",
              {
                "Ref": "MyBucketF68F3FF0",
              },
              "","Key":"dummy.json"},"physicalResourceId":{"id":"s3:://",
              {
                "Ref": "MyBucketF68F3FF0",
              },
              "/dummy.json"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": false,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": {
          "Fn::Join": [
            "",
            [
              "{"service":"S3","action":"putObject","parameters":{"Body":"{\\"dummy\\":\\"data\\"}","Bucket":"",
              {
                "Ref": "MyBucketF68F3FF0",
              },
              "","Key":"dummy.json"},"physicalResourceId":{"id":"s3:://",
              {
                "Ref": "MyBucketF68F3FF0",
              },
              "/dummy.json"}}",
            ],
          ],
        },
        "salt": 1730494729270,
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "MyS3FileResourceCustomResourcePolicy761A62A0": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MyBucketF68F3FF0",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MyBucketF68F3FF0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyS3FileResourceCustomResourcePolicy761A62A0",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
